File: assignments/assignment-1-solution/src/CrazyEights/CardDeck/Deck.cs
using CrazyEights.Cards;

namespace CrazyEights.CardDeck;

public class Deck
{
    private readonly Stack<ICard> _cards;

    public Deck(IEnumerable<ICard> cards)
    {
        ArgumentNullException.ThrowIfNull(cards);
        _cards = new Stack<ICard>(cards);
    }

    public int CardCount => _cards.Count;

    public bool IsEmpty => _cards.Count == 0;

    public ICard DrawCard()
    {
        if (IsEmpty) throw new InvalidOperationException("Deck is empty.");

        return _cards.Pop();
    }
}

File: assignments/assignment-1-solution/src-java/src/main/java/edu/kennesaw/crazy8s/carddeck/Deck.java
package edu.kennesaw.crazy8s.carddeck;

import edu.kennesaw.crazy8s.cards.Card;
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.List;

/**
 * Represents a stack-based deck of cards to draw from.
 */
public class Deck {
    private final Deque<Card> cards = new ArrayDeque<>();

    public Deck(List<Card> cards) {
        if (cards == null) {
            throw new IllegalArgumentException("Cards cannot be null.");
        }
        for (Card card : cards) {
            this.cards.push(card);
        }
    }

    public int getCardCount() {
        return cards.size();
    }

    public boolean isEmpty() {
        return cards.isEmpty();
    }

    public Card drawCard() {
        if (isEmpty()) {
            throw new IllegalStateException("Deck is empty.");
        }

        return cards.pop();
    }
}
